src/main/java/org/truffleruby/core/format/TruncateStringNode.java

```diff
@@ -0,0 +1,24 @@
+package org.truffleruby.core.format;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import org.truffleruby.language.library.RubyStringLibrary;
+import org.truffleruby.core.rope.Rope;
+import java.util.Arrays;
+
+public class TruncateStringNode extends FormatNode {
+
+    private final FormatNode value;
+    private final Integer size;
+
+    public TruncateStringNode(FormatNode value, Integer size) {
+        this.value = value;
+        this.size = size;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        byte[] val = (byte[]) this.value.execute(frame);
```

Comment:
All I have at this layer is `Object` coming from `this.value.execute(frame)`.
How do I even go from `Object` to `String`/`RubyString`/`Rope` and truncate it properly?

Reply:
Yes, the typing here isn't very strong! Possibly `FormatNode` should be split into several more specific nodes with useful types. For now you just have to know what types you're getting and cast, as you're doing.

It looks like by the time we truncate we have already lowered to an encoding-less raw `byte[]` representation. I was going to suggest therefore that the precision looks like the number of *bytes* to write, but a couple of tests show unfortunately this isn't the case.

```
format('%.3s', 'foo'.encode('utf-16le')).bytesize # => 6
'😊'.bytesize #=> 4
format('%.3s', '😊').bytesize # => 4
```

So it does look like it should be truncating by characters. So we've got ourselves into a bit of a pickle here! We're trying to do something after we threw away the info we needed for it!

I think the MRI logic is around here and the `FPREC` flag.

https://github.com/ruby/ruby/blob/8a4472fb6d2df0f6407cef24df6a038be90d1462/sprintf.c#L500

The truncate probably needs to be inserted earlier in the pipeline, before the encoding gets thrown away.

We could possibly do with some multibyte string specs like I'm demonstrating above.

@eregon do you have any opinions on this?

Reply:
https://github.com/oracle/truffleruby/compare/master...Shopify:string-format-precision-v2

Here's another way to implement it that I came up with. It's operating with RubyString and it should preserve the encoding.

Reply:
Yes that's probably it - raise the truncate operation higher.

Reply:
That approach looks good (see my other comment).
Yes, specs are always good.
This PR should add an example with a multibyte String that actually truncates some characters to make sure that works correctly.